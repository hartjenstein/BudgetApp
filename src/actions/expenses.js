import uuid from 'uuid';
import database from '../firebase/firebase';

// Action generators for expenses reducer //

///* flow without firebase:
// component calls actoin generator
// action generator returns object
// component dispatches object
// redux store changes

///*flow with firebase
// component calls action generator
// action generator returns function
// component dispatches function (?)
// function runs (has the ability to dispatch other actions and do whatever it wants)
//redux by default does not allow to dispatch functions. Thats what we use thunk for

//ADD_EXPENSE
//action generator without firebase
/* export const addExpense = ({ 
    description = '', 
    note = '', 
    amount = 0, 
    createdAt = 0 
} = {}) =>  ({
    type: 'ADD_EXPENSE',
    expense: {
        id: uuid(),
        description,
        note,
        amount,
        createdAt
}
}); */
// the ADD_EXPENSE action generator (above) is simplefied and we add the  async action gen. startAddExpense, as firebase database only takes objects and 
// the redux store state only accepts arrays.
export const addExpense = (expense) => ({
	type: 'ADD_EXPENSE',
	expense
});

export const startAddExpense = (expenseData = {}) => {
  // passing dispatch as an arguement, this is happening internally with thunk middleware and redux! 
	return (dispatch) => {
		const {
			//moving defaults here from ADD_EXPENSE, destructuring inside function instead of in the arguements to make it readable more easily
			description = '',
				note = '',
				amount = 0,
				createdAt = 0
		} = expenseData;
		const expense = { description, note, amount, createdAt };
		// saving data to firebase, then dispatching the action
		// asyncronus action generator - promise comes back from firebase 
		//and gets called with ref as an arguement - we get the ref object returned from our call to the database
		// ref.key contains the id firebase created for data set we've just pushed
		// 
		return database.ref('expenses')
			.push(expense)
			.then((ref) => {
				dispatch(addExpense({
					id: ref.key,
					...expense
				}));
		});
	};
};

// REMOVE_EXPENSE 
export const removeExpense = ({ id } = {}) => ({
	type: 'REMOVE_EXPENSE',
	id
});

// EDIT_EXPENSE 
export const editExpense = (id, updates) => ({
	type: 'EDIT_EXPENSE',
	id,
	updates
});

export const startEditExpense = (id, updates) => {
  return (dispatch) => {
    return database.ref(`expenses/${id}`).update({
      ...updates
    }).then(() => {
      dispatch(editExpense(id, updates))
    });
  };
}; 

// SET_EXPENSES
export const setExpenses = (expenses) => ({
  type: 'SET_EXPENSES',
  expenses
});

//Async function to handle firebase set data
export const startSetExpenses = () => {
  // dispatch arguement is provided internally from thunk middleware 
 return (dispatch) => {
    return database.ref('expenses').once('value').then((snapshot) => {
      const expenses = [];
      snapshot.forEach((childSnapshot) => {
        expenses.push({
          id: childSnapshot.key,
          ...childSnapshot.val()
        });
      });
      dispatch(setExpenses(expenses));
    });
  }
};

export const startRemoveExpense = ({ id } = {}) => {
  return (dispatch) => { // dispatch arguement gets passed in by redux library / redux thunk middleware
    return database.ref(`expenses/${id}`).remove().then(() => {
      dispatch(removeExpense({ id }));
    });
  };
};