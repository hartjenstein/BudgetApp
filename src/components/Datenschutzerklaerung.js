import React from 'react';
import Header from './Header';

const Datenschutzerklaerung = () => (
  <div>
  <Header />
    <div className="content-container content">
      <h1>Datenschutzerklärung:</h1>


      <h2>Verantwortliche Stelle</h2>
      
      <p>Verantwortliche Stelle für die Erhebung, Verarbeitung und Nutzung Ihrer personenbezogenen Daten im Sinne des Bundesdatenschutzgesetzes (BDSG) ist
      
      N Hartjenstein, Richardstr. 101b, 12043 Berlin
      
      Sofern Sie der Erhebung, Verarbeitung oder Nutzung Ihrer Daten durch N Hartjenstein gemäß den datenschutzrechtlichen Bestimmungen insgesamt oder für einzelne Maßnahmen widersprechen möchten, richten Sie Ihren Widerspruch per Brief an folgende Kontaktdaten:
      
      Richardstr. 101b, 12043 Berlin
      
      Wir legen größten Wert auf den Schutz Ihrer Daten und die Wahrung Ihrer Privatsphäre. Nachstehend informieren wir Sie deshalb über die Erhebung und Verwendung persönlicher Daten bei Nutzung unserer Webseite.
      </p>
      
      <h2>Erheben, Speichern und Verwenden personenbezogener Daten</h2> 
        
      <p>Bei dem Besuch unserer Website erheben, verwenden und speichern wir Ihre personenbezogenen Daten. Personenbezogene Daten sind nach § 3 Abs. 1 BDSG Einzelangaben über persönliche oder sachliche Verhältnisse einer bestimmten oder bestimmbaren natürlichen Person (Betroffener).
        
        
        Datenerhebung unter strenger Beachtung der Gesetze
        
        Bei der Erhebung, dem Verwenden und Speichern von personenbezogenen Daten beachten wir die Vorschriften des Bundesdatenschutzgesetzes, des Telemediengesetzes und ggf. weiterer anwendbarer Datenschutzbestimmungen. Alle Kundendaten werden unter Beachtung der einschlägigen Vorschriften des Bundesdatenschutzgesetzes (BDSG) und des Telemediengesetzes (TMG) von uns gespeichert und verarbeitet.
        </p>
        <h2>Cookies</h2>
        
        <p>Unsere Internetseite verwendet an mehreren Stellen so genannte Cookies. „Cookies“ sind kleine Textdateien, die auf Ihrem Rechner abgelegt werden und die Ihr Browser speichert. Sie dienen dazu, unser Angebot nutzerfreundlicher, effektiver und sicherer zu machen. Des Weiteren ermöglichen Cookies unseren Systemen, Ihren Browser zu erkennen und Ihnen Services anzubieten. Cookies enthalten keine Personenbezogenen Daten. Sie haben die Möglichkeit, das Abspeichern von Cookies auf Ihrem Rechner durch entsprechende Einstellungen in Ihrem Browser zu verhindern. Hierdurch könnte allerdings der Funktionsumfang unseres Angebotes eingeschränkt werden.
        </p>
        
        <h2>Google-Analytics</h2>
        
        <p>Diese Website benutzt Google Analytics, einen Webanalysedienst der Google Inc. („Google“). Google Analytics verwendet sog. „Cookies“, Textdateien, die auf Ihrem Computer gespeichert werden und die eine Analyse der Benutzung der Website durch Sie ermöglichen. Die durch den Cookie erzeugten Informationen über Ihre Benutzung dieser Website werden in der Regel an einen Server von Google in den USA übertragen und dort gespeichert.
        Im Falle der Aktivierung der IP-Anonymisierung auf dieser Webseite wird Ihre IP-Adresse von Google jedoch innerhalb von Mitgliedstaaten der Europäischen Union oder in anderen Vertragsstaaten des Abkommens über den Europäischen Wirtschaftsraum zuvor gekürzt. Nur in Ausnahmefällen wird die volle IP-Adresse an einen Server von Google in den USA übertragen und dort gekürzt. Die IP-Anonymisierung ist auf dieser Website aktiv. Im Auftrag des Betreibers dieser Website wird Google diese Informationen benutzen, um Ihre Nutzung der Website auszuwerten, um Reports über die Websiteaktivitäten zusammenzustellen und um weitere mit der Websitenutzung und der Internetnutzung verbundene Dienstleistungen gegenüber dem Websitebetreiber zu erbringen.
        Die im Rahmen von Google Analytics von Ihrem Browser übermittelte IP-Adresse wird nicht mit anderen Daten von Google zusammengeführt. Sie können die Speicherung der Cookies durch eine entsprechende Einstellung Ihrer Browser-Software verhindern; wir weisen Sie jedoch darauf hin, dass Sie in diesem Fall gegebenenfalls nicht sämtliche Funktionen dieser Website vollumfänglich werden nutzen können. Sie können darüber hinaus die Erfassung der durch das Cookie erzeugten und auf Ihre Nutzung der Website bezogenen Daten (inkl. Ihrer IP-Adresse) an Google sowie die Verarbeitung dieser Daten durch Google verhindern, indem Sie das unter dem folgenden Link verfügbare Browser-Plugin herunterladen und installieren: http://tools.google.com/dlpage/gaoptout?hl=de
        </p>
        
        
        <h2>Auskunft</h2>

        <p>
        Sie haben jederzeit das Recht auf unentgeltliche Auskunft über Ihre gespeicherten Daten sowie das Recht auf Berichtigung, Löschung bzw. Sperrung. Wenden Sie sich hierzu bitte an die im Impressum angegebene Kontaktadresse. Wir stehen Ihnen jederzeit gern für weitergehende Fragen zu unseren Hinweisen zum Datenschutz und zur Verarbeitung Ihrer persönlichen Daten zur Verfügung.
        

        </p>
    </div>
    </div>
);

export default Datenschutzerklaerung;