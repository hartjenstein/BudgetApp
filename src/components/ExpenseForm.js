import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';


//const date = new Date();
//moment example
const now = moment();
console.log(now.format('MMM Do, YYYY'));

// * using local state to keep track of the changes the user makes in the form
// only if user submits we send it of to redux
// * why did we switch back to the constructor function? We want to 
//populate the state with exitsing state which we get passed down from the edit expense page 
// via props. If we dont set up the contructor function there is no way to access the props! 
export default class ExpenseForm extends React.Component {
    constructor(props) {
        super(props);

        this.state = {
            description: props.expense ? props.expense.description : '',
            note: props.expense ? props.expense.note : '',
            amount: props.expense ? (props.expense.amount / 100).toString() : '',
            createdAt: props.expense ? moment(props.expense.createAt) : moment(),
            calenderFocused: false,
            error: ''
        };
    };

    onDescriptionChange = e => {
        const description = e.target.value;
        this.setState(() => ({ description })); 
    };

    onNoteChange = e => {
        const note = e.target.value;
        this.setState(() => ({ note }));    
    };

    onAmountChange = e => {
        const amount = e.target.value;
        if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
            this.setState(() => ({ amount }));
        } 
    };

    onDateChange = (createdAt) => {
        if (createdAt) {
        this.setState(() => ({ createdAt }));
        }
    };

    onFocusChange = ({ focused }) => {
        this.setState(() => ({ calenderFocused: focused }));
    };

    onSubmit = e => {
        e.preventDefault();

        if (!this.state.description ||Â !this.state.amount) {
            this.setState(() => ({ error: 'Please enter a description and a amount'}));
        } else {
            this.setState(() => ({ error: '' }));
            this.props.onSubmit({ 
                description: this.state.description,
                amount: parseFloat(this.state.amount, 10) * 100,
                // valueOF() -> momentJS method to get the timestamp back
                createdAt: this.state.createdAt.valueOf(),
                note: this.state.note
            })
        }
    };

    render() {
        return (
          <form className="form" onSubmit={this.onSubmit}>
          { this.state.error && <p className="form__error" >{this.state.error}</p>}
              <input type="text"
                  className="input--text"
                  placeholder="Description"
                  autoFocus
                  //ontrolled components for single point of truth
                  value={this.state.description}
                  onChange={this.onDescriptionChange}
              />
              <input type="text" 
                  className="input--text" 
                  placeholder="Amount"
                  value={this.state.amount}
                  onChange={this.onAmountChange}
              />
              <SingleDatePicker 
                  //required
                  date={this.state.createdAt}
                  onDateChange={this.onDateChange}
                  focused={this.state.calenderFocused}
                  onFocusChange={this.onFocusChange}
                  // custom
                  numberOfMonths={1}
                  isOutsideRange={() => false}
              />
              <textarea 
              className="input--textarea"
                  placeholder="Add a Note to your expense"
                  value={this.state.note}
                  onChange={this.onNoteChange}
                  >
              </textarea>
              <div>
                <button className="button">Save Expense</button>
              </div>
              
          </form>
        )
    }
}