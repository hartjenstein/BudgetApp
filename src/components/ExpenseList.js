import React from 'react';
import { connect } from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import selectExpenses from '../selectors/expenses';

//exporting the unconnected function for testing with enzyme
export const ExpenseList = (props) => (
    <div className="content-container">
      <div className="list__header">
        <div className="show-for-mobile">Expenses</div>
        <div className="show-for-desktop">Expense</div>
        <div className="show-for-desktop">Amount</div>
      </div>
        <div className="list__body">
          {
              props.expenses.length === 0 ? (
                <div className="list__item list__item--message">
                  <span>No expenses</span>
                </div>
      
              ) : props.expenses.map((expense) => (
                  <ExpenseListItem 
                  key={expense.id}
                  {...expense}
                  />
          ))}
        </div>
    </div>
);

//Connecting the component to the redux store with a higher order component (hoc)
// Connect returns another function which 
// creates a higher order component in order to connect to the store
// the stores state gets passed in and we return the props we need out of the state

const mapStateToProps = (state) => {
    return {
        expenses: selectExpenses(state.expenses, state.filters)
    }
};

export default connect(mapStateToProps)(ExpenseList);
