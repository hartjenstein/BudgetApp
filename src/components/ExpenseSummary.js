import React from 'react';
import { connect } from 'react-redux';
import selectExpensesTotal from '../selectors/expenses-total';
import selectExpenses from '../selectors/expenses';
import numeral from 'numeral';

export class ExpenseSummary extends React.Component {

    expenseTotal() {
        const totalExpense = selectExpensesTotal(this.props.expenses);
        return numeral( totalExpense / 100).format('$0,0.00');
    }
    render() {
        return (
            <div>
                <h3>
                
                   You have {this.props.expenses.length} {this.props.expenses.length === 1 ? 'Expense' : 'Expenses' } with a total of {this.expenseTotal()}
                    
                </h3>
            </div>
        )
    }
};


const mapStateToProps = (state) => {
    return {
        expenses: selectExpenses(state.expenses, state.filters)
    }
};

export default connect(mapStateToProps)(ExpenseSummary);