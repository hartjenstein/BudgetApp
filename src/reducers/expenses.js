// Expenses Reducer

const expensesReducerDefaultState = [];
export default (state = expensesReducerDefaultState, action) => {
    switch (action.type) {
        case 'ADD_EXPENSE': 
        //use concat instead of push to avoid mutating the array
        // here we use array destructuring and the rest operator
          return [...state, action.expense];
        case 'REMOVE_EXPENSE': 
            return state.filter(({ id }) => id !== action.id);
        case 'EDIT_EXPENSE': 
            return state.map((expense) => {
                if (expense.id === action.id) {
                    return {...expense,
                        ...action.updates
                    };
                } else {
                    return expense
                }
            });
        case 'SET_EXPENSES': 
            return action.expenses;
        default: 
            return state;
    }
};
