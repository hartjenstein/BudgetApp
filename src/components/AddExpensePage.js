import React from 'react';
import ExpenseForm from './ExpenseForm';
import { connect } from 'react-redux';
import { startAddExpense } from '../actions/expenses.js'

// adding export addoitionally to class to make it testable without connect();
export class AddExpensePage extends React.Component {
  onSubmit = (expense) => {
    this.props.startAddExpense(expense);
    this.props.history.push('/');
  }
  render() {
    return (
      <div>
        <div className="page-header">
          <div className="content-container">
            <h1 className="page-header__title">Add Expense</h1>
          </div>
        </div>
        <div className="content-container">
          <ExpenseForm 
          onSubmit={this.onSubmit}
          />
        </div>
      
    </div>
    );
  }
}
/// replaced by a class with onSubmit() and mapDispatchToProps() to make it testable 
// - the problem is that the call to dispatch (down below) is referencing 
/* const AddExpensePage = (props) => (
    <div>
      <h1>Add Expense</h1>
      <ExpenseForm 
      onSubmit={(expense) => { */
        // abstracted away (down below) for testing purposes and replaced with a call to onSubmit()
        // props.dispatch(addExpense(expense));
      /*  props.history.push('/');
      }}/>
    </div>
  ); */
///
    //mapDispatchToProps in order to make this component testable with jest
  // we set up the function that connect() takes as a second arguement
  // with mapDispatchProps we pass the call to dispatch and make it testable this way
  //( mapDispatchProps -> a way to return dispatcher functions, abstracting them away from the component itself)
  const mapDispatchToProps = (dispatch) => ({ 
    startAddExpense: (expense) => dispatch(startAddExpense(expense))
  });
    
  
 
  // arguement 
  // if you only need the dispatch method you can call connect without argument
  export default connect(undefined, mapDispatchToProps)(AddExpensePage);