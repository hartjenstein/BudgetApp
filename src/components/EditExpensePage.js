import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { editExpense, removeExpense } from '../actions/expenses';
// Refactor to class based components so we can convert inline functions (meaning in JSX defined) to methods, 
// so that they dont have to be redefined everytime the component gets rendered (for testing purposes )

// there are special props that are passed down by react router like the history object
export class EditExpensePage extends React.Component {
  handleOnSubmit = (expense) => {
    this.props.editExpense(this.props.expense.id, expense);
    this.props.history.push('/');
  };

  handleOnClick= () => {
    this.props.removeExpense({id: this.props.expense.id})
    this.props.history.push('/');
  };

  render() {
    return (
      <div>
        <ExpenseForm 
            expense={this.props.expense}
            onSubmit={this.handleOnSubmit}
        />
        <button 
          onClick={this.handleOnClick}
        >Remove</button>
      </div>
    );
  }
};

// we are searching the state object in the store for the id we get from the props that are passed down by the router 
// we can pass state and props as arguments to the higher order component (hoc) in order to do that
const mapStateToProps = (state, props) => {
  return {
    // we generate a new prop called expense and add it on to props
    expense: state.expenses.find((expense) => expense.id === props.match.params.id)
  };
};

const mapDispatchToProps = (dispatch, props) => (
  { editExpense: (id, expense) => dispatch(editExpense(id, expense)),
    removeExpense: (data) => dispatch(removeExpense(data))
  });

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);