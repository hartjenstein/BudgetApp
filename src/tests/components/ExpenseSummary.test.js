import React from 'react';
import Numeral from 'numeral';
import { shallow } from 'enzyme';
import { ExpenseSummary } from '../../components/ExpenseSummary';
import expenses from '../fixtures/expenses';


test('should render correctly with 1 expense', () => {
  const wrapper = shallow(<ExpenseSummary expenses={[expenses[1]]}/>);
  expect(wrapper).toMatchSnapshot();
});

test('should render correctly with multiple expenses', () => {
  const wrapper = shallow(<ExpenseSummary expenses={expenses}/>);
  expect(wrapper).toMatchSnapshot();
});