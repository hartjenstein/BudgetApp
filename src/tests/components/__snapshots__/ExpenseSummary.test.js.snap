// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly with 1 expense 1`] = `
<div>
  <h3>
    You have 
    1
     
    expense
     with a total of 
    $0.14
  </h3>
</div>
`;

exports[`should render correctly with multiple expenses 1`] = `
<div>
  <h3>
    You have 
    3
     
    expenses
     with a total of 
    $0.41
  </h3>
</div>
`;
